compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/PrefetchSharedMemoryCopy.cpp

```diff
@@ -176,9 +176,54 @@ class LoopPrefetcher {
     });
   }
 
+  void getValueDependenciesForIf(scf::IfOp ifOp,
+                                 DenseSet<Operation *> &readDependencies,
+                                 DenseSet<Operation *> &writeDependencies) {
+    // scf.if with results should be supported directly through the usual
+    // handling so bail-out here in that case
+    if (ifOp->getNumResults() != 0)
+      return;
+    bool hasGlobalRead = false;
+    bool hasSharedWrite = false;
+    // Else region not yet supported.
+    if (!ifOp.getElseRegion().empty()) {
+      return;
+    }
+    ifOp->walk([&](Operation *op) {
+      if (auto readOp = dyn_cast<vector::TransferReadOp>(op)) {
+        auto sourceType = dyn_cast<MemRefType>(readOp.getBase().getType());
+        if (hasGlobalMemoryAddressSpace(sourceType)) {
+          hasGlobalRead = true;
+        }
+      }
+      if (auto writeOp = dyn_cast<vector::TransferWriteOp>(op)) {
+        auto dstType = dyn_cast<MemRefType>(writeOp.getBase().getType());
+        if (hasSharedMemoryAddressSpace(dstType)) {
+          hasSharedWrite = true;
+        }
+      }
+    });
+    // if op has both read and write stages and hence we cannot do prefetching.
+    if (hasGlobalRead && hasSharedWrite) {
+      return;
+    }
+    if (hasSharedWrite) {
+      getValueDependencies(ifOp, writeDependencies, /*noTransferReads=*/true);
+      return;
+    }
+    // By default we can mark the if op as read stage.
```

Comment:
I think we'd want to differentiate between 1) has global read 2) no global read nor shared write.

If this is 1) then categorizing this into read dependency is correct.

If this is 2) and there's no global load nor shared write, then we are almost sure that this belong to the read stage. However, what if in the future the prefetch come to deal with the scenario that `scf.if` compute some result that write stage will depend on? We may want to bail out on that scenario just to be defensive.

Reply:
yeah that all needs private memory handling (the TODO comment) but I agree with you currently I can just bail-out for it, this means the conv padding case will just bail-out though which is okay for now.